src/methods/large_shuffle.jl

```diff
@@ -0,0 +1,169 @@
+    T = eltype(x)
+    s = Vector{T}(undef, 0)
+    sizehint!(s, L)
+
+    startinds = [1; cs .+ 1]
+    @inbounds for i in draw_order
+        inds = startinds[i]:startinds[i]+Ls[i]-1
+        append!(s, xrot[inds])
+    end
+
+    return s
+end
+
+#########################################################################
+# CycleShuffle
+#########################################################################
+"""
+    CycleShuffle(firstlast::Bool = false; n::Int = 7, σ = 0.5) <: Surrogate
+
+Cycle shuffled surrogates[^Theiler1995] that identify successive local peaks in the data and shuffle the
+cycles inbetween the peaks. Similar with [`BlockShuffle`](@ref), but here
+the "blocks" are defined as follows:
+1. The timeseries is smoothened via convolution with Gaussian (`DSP.gaussian(n, σ)`).
+2. Local maxima of the smoothened signal define the peaks, and thus the blocks in between them.
+3. The first and last index of timeseries can never be peaks and thus signals that
+   should have peaks very close to start or end may not perform well. In addition,
+   points before the first or after the last peak are never shuffled.
+3. The defined blocks are randomly shuffled as in [`BlockShuffle`](@ref).
+
+CSS are used to test the null hypothesis that the signal is generated by a periodic
+oscillator with no dynamical correlation between cycles,
+i.e. the evolution of cycles is not deterministic.
+
+See also [`PseudoPeriodic`](@ref).
+
+[^Theiler1995]: J. Theiler, On the evidence for low-dimensional chaos in an epileptic electroencephalogram, [Phys. Lett. A 196](https://doi.org/10.1016/0375-9601(94)00856-K)
+"""
+struct CycleShuffle{T <: AbstractFloat} <: Surrogate
+    n::Int
+    σ::T
+end
+CycleShuffle(n = 7, σ = 0.5) = CycleShuffle{typeof(σ)}(n, σ)
+
+function surrogenerator(x::AbstractVector, cs::CycleShuffle)
+    n, N = cs.n, length(x)
+    g = DSP.gaussian(n, cs.σ)
+    smooth = DSP.conv(x, g)
+    r = length(smooth) - N
+    smooth = iseven(r) ? smooth[r÷2+1:end-r÷2] : smooth[r÷2+1:end-r÷2-1]
+    peaks = findall(i -> smooth[i-1] < smooth[i] && smooth[i] > smooth[i+1], 2:N-1)
+    blocks = [collect(peaks[i]:peaks[i+1]-1) for i in 1:length(peaks)-1]
+    init =  (blocks = blocks, s = copy(x), peak1 = peaks[1])
+    SurrogateGenerator(cs, x, init)
+end
+
+function (sg::SurrogateGenerator{<:CycleShuffle})()
+    blocks, s, peak1 = sg.init
+    x = sg.x
+    shuffle!(blocks)
+    i = peak1
+    for b in blocks
+        s[(0:length(b)-1) .+ i] .= @view x[b]
+        i += length(b)
+    end
+    return s
+end
+
+#########################################################################
+# Timeshift
+#########################################################################
+"""
+    CircShift(n) <: Surrogate
+Surrogates that are circularly shifted versions of the original timeseries.
+
+`n` can be an integer (meaning to shift for `n` indices), or any vector of integers,
+which means that each surrogate is shifted by a random entry of `n`.
```

Comment:
I had to read this a few times before I understood it.  Perhaps it would be clearer to say "... which means that each surrogate is shifted by an integer, selected randomly among the entries in `n`"?